### PR TITLE
add optional compressed RGB565 MARLIN logos

### PR DESCRIPTION
**Pull request by ellensp**
### Description

Adding cutting edge 1983's technology into Marllin.

On Marlin controlled TFT's the Marlin bootscreen can take up a lot of flash space (even with `BOOT_MARLIN_LOGO_SMALL`)

```
uint16_t marlin_logo_195x59x16[11505]     ie 23010 bytes  (22K)
uint16_t marlin_logo_320x240x16[76800]   ie 153600 bytes (150K)
uint16_t marlin_logo_480x320x16[153600] ie 307200 bytes (300K)
```

This PR allows the user to use the same screen images but the data is compressed with a simple run length encoding, reducing the flash space required by approximately half. 

```
uint8_t marlin_logo_195x59x16_rle16[13953]        13953 bytes (13.6K)
uint8_t marlin_logo_320x240x16_rle16[73695]      73695 bytes (71.9K)
uint8_t marlin_logo_480x320x16_rle16[123126]  123126 bytes (120.2K)
```

Added a new option to Configuration_adv.h   ~~`BOOT_MARLIN_LOGO_USES_RLE16`~~ replaced by `COMPACT_MARLIN_BOOT_LOGO`

A new script has been added buildroot/share/scripts/rel16_compress_cpp_image_data.py that creates the required compressed data from the original cpp data file.
Update:  the online rgb565 converter https://marlinfw.org/tools/rgb565/converter.html is getting an update to also generate compressed data https://github.com/MarlinFirmware/MarlinDocumentation/pull/512

How does it works

(I don't know if there is a official name for this implementation, it is normally 8 bit count and 8 bit data, but a 8 bit count and 16 bit data make much better sense here, thus the name I have given it RLE16) 
 
Updated:
    # RLE16 (run length 16) encoding
    # Convert data from from raw RGB565 to a simple run-length-encoded format for each word of data.
    # - Each sequence begins with a count byte N.
    #   - If the high bit is set in N the run contains N & 0x7F + 1 unique words.
    #   - Otherwise it repeats the following word N + 1 times.
    # - Each RGB565 word is stored in MSB / LSB order.

So if you have for eg a run of 100 RGB565 0x188B instead of taking 200 bytes to store this it now is stored in 3 bytes,
Of course there is no such thing as a free lunch. Unique pixels used to take 2 bytes, but now the block has a 1 byte header, making this section slightly larger then its uncompressed counterpart. 

There is also a tiny code addition required to decompress the images.

### Requirements

`TFT_COLOR_UI`

### Benefits

Smaller firmware

### Configurations

Test Configuration

```cpp
#define MOTHERBOARD BOARD_SIMULATED
#define TFT_COLOR_UI
#define TFT_GENERIC
#define TFT_INTERFACE_SPI
#define TOUCH_SCREEN
  // One of these
  //#define TFT_RES_320x240
  //#define TFT_RES_480x272
  //#define TFT_RES_480x320
  //#define TFT_RES_1024x600
```

### Related Issues
https://github.com/MarlinFirmware/Marlin/issues/25963

### Additional notes 
Also fixed a bug with `TFT_RES_1024x600` using 480x320 bootscreen, this did not display correctly compressed or otherwise. 
Moved SITE_URL down on 480x272 boot screen, was in wrong Y position.

### CODE REVIEW

**Comment by thisiskeithb** [comment]
> TFT_RES_1024x600 uses 480x320 bootscreen, this does not display correctly compressed or otherwise. This is a bug for another day

Since only the Biqu BX runs this resolution, I figured I'd look into that/create one with `BOOT_MARLIN_LOGO_SMALL` disabled, but it bootloops on current `bugfix-2.1.x`. 🤷

I'll have to track down when that first started apparently.

**Comment by ellensp** [comment]
@thisiskeithb fixed the issue with 480x320 bootscreen on 1024x600 display.  (on simulator)

**Comment by thisiskeithb** [comment]
> fixed the issue with 480x320 bootscreen on 1024x600 display. (on simulator)

Now it'll break if you try to use one that is 1024x600 😄

**Comment by ellensp** [comment]
yes it will, luckily there is no such boot screen in marlin :)

**Comment by thisiskeithb** [comment]
> yes it will, luckily there is no such boot screen in marlin :)

...Yet! 😄 That's what I was going to create to test this PR until I discovered the bootloop issue on the BX.

Seems "🎨 Lowercase methods, functions, data members (https://github.com/MarlinFirmware/Marlin/pull/25939)" / https://github.com/MarlinFirmware/Marlin/commit/c3694f1c038fa20ad749118b0a1c7aa88e3dd5f2 is the culprit.

**Comment by thisiskeithb** [comment]
Here's an upscaled 1024x600x16 bootscreen + required code changes for `TFT_COLOR_UI`: https://github.com/thisiskeithb/Marlin/commit/4652d211ba1bec88c5752e9578ec7dfa1a03a628

> marlin_logo_1024x600x16[614400] = 1,228,800 bytes (1.2288MB)

This is absolutely **huge** and would require boards with at least 2mb of flash (before this PR):

```prolog
RAM:   [=         ]   6.2% (used 65144 bytes from 1048576 bytes)
Flash: [=======   ]  73.8% (used 1548420 bytes from 2097152 bytes)
```

With `BOOT_MARLIN_LOGO_SMALL` enabled (before this PR):

```prolog
RAM:   [=         ]   6.2% (used 65144 bytes from 1048576 bytes)
Flash: [==        ]  16.3% (used 342628 bytes from 2097152 bytes)
```

Even at half the size achieved through compression, I'm not sure it's worth eating up that much space and would recommend sticking with centering the 480x320 bootscreen for 1024x600 when `BOOT_MARLIN_LOGO_SMALL` is disabled.

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -1547,6 +1547,9 @@
     #if ANY(HAS_MARLINUI_U8GLIB, TFT_COLOR_UI)
       #define BOOT_MARLIN_LOGO_SMALL      // Show a smaller Marlin logo on the Boot Screen (saving lots of flash)
     #endif
+    #if ENABLED(TFT_COLOR_UI)
+      //#define BOOT_MARLIN_LOGO_USES_RLE16 // Uses less flash space.
+    #endif
```
Comment (thisiskeithb) [inline Marlin/Configuration_adv.h]: I'd move this to the existing `#if ENABLED(TFT_COLOR_UI)` block in `Configuration.h`

https://github.com/MarlinFirmware/Marlin/blob/60f591f3b6e06fdef903e10f81ac96f65e30d78e/Marlin/Configuration.h#L3288-L3299

**Comment by thinkyhead** [comment]
The RLE scheme can be improved so that it never produces run-lengths of 1 and can also produce runs of distinct values when this results in a smaller output. I suggest that we reduce the size of runs to a maximum of 128 and use the high bit to indicate that the run is unique and not repeated. I'll code that up in the Python script and then adapt the RLE decoder in C++ to decompress the new scheme.

There is also a potential optimization where each run has a size. Ultimately, the best scheme would be to compress with Huffmann encoding but this requires extra logic and a table of values sorted by frequency, so that can wait until further experimentation.

**Comment by thinkyhead** [comment]
With these changes to RLE encoding we get ~~significantly~~ slightly better compression:
- 480x320 (307200): Was 123126 (60%), now 104852 (66%)
- 320x240 (153600): Was 73695 (52%), now 62287 (60%)
- 195x59 (23010): Was 13953 (40%), now 11753 (49%)

File: Marlin/src/inc/Changes.h [hunk 2]
```diff
@@ -659,6 +659,8 @@
   #error "Z3_USE_ENDSTOP is obsolete. Instead set Z2_STOP_PIN directly. (e.g., 'Z3_USE_ENDSTOP _ZMAX_' becomes 'Z3_STOP_PIN Z_MAX_PIN')"
 #elif defined(Z4_USE_ENDSTOP)
   #error "Z4_USE_ENDSTOP is obsolete. Instead set Z4_STOP_PIN directly. (e.g., 'Z4_USE_ENDSTOP _ZMAX_' becomes 'Z4_STOP_PIN Z_MAX_PIN')"
+#elif defined(BOOT_MARLIN_LOGO_USES_RLE16)
+  #error "BOOT_MARLIN_LOGO_USES_RLE16 is now COMPACT_MARLIN_BOOT_LOGO."
```
Comment (thisiskeithb) [inline Marlin/src/inc/Changes.h]: `BOOT_MARLIN_LOGO_USES_RLE16` was never an option before this PR, so it doesn't need to be added here.

**Comment by thinkyhead** [comment]
> Seems #25939 is the culprit.

Have you been able to trace the reboot down to the specific change that is causing the problem yet? I haven't flashed my BX in a while, but I can do that if you are otherwise occupied.

**Comment by thisiskeithb** [comment]
> Have you been able to trace the reboot down to the specific change that is causing the problem yet? I haven't flashed my BX in a while, but I can do that if you are otherwise occupied.

There's too much changed there, so reverting to the prior commit was the fix for now.

**Comment by ellensp** [comment]
I was deliberately trying to keep this simple... no weird bit level stuff, and definitely not Huffmann 

But I cant argue with the better results of using runs of compressed and uncompressed data.

**Comment by ellensp** [comment]
Currently with COMPACT_MARLIN_BOOT_LOGO enabled the image is corrupted  
Screenshot from 2023-06-24 18-13-44

now fixed. after 55aa442 in sim

**Comment by ellensp** [comment]
This has fixed it.

**Comment by ellensp** [comment]
Tested this on BOARD_MKS_ROBIN_NANO_V3 with MKS_TS35_V2_0  (480x320)
//#define BOOT_MARLIN_LOGO_SMALL and   #define COMPACT_MARLIN_BOOT_LOGO

This works as expected, boots and displays the fullsize Marlin boot screen

With COMPACT_MARLIN_BOOT_LOGO
Flash: [===       ]  28.3% (used 296976 bytes from 1048576 bytes)
without COMPACT_MARLIN_BOOT_LOGO
Flash: [=====     ]  47.6% (used 499136 bytes from 1048576 bytes)

**Comment by ellensp** [comment]
lol, just found marlin has rle_compress and rle_uncompress  in Marlin/src/lcd/extui/mks_ui/SPIFlashStorage.cpp
I was unaware there was already some rle in the code base

**Comment by EvilGremlin** [comment]
yep, there are things in LVGL_UI that need to be moved outside for global access

**Comment by jmz52** [comment]
FYI - just found out that the [lcd-image-converter](https://sourceforge.net/projects/lcd-image-converter/) tool I am using to convert images supports RLE compression for generated "C" data

**Comment by thisiskeithb** [comment]
> FYI - just found out that the [lcd-image-converter](https://sourceforge.net/projects/lcd-image-converter/) tool I am using to convert images supports RLE compression for generated "C" data

Ours needs to be updated to support it (and proper parsing of RLE data when this PR is merged): https://marlinfw.org/tools/rgb565/converter.html

**Comment by thinkyhead** [comment]
- Compressed RLE16 boot screens are now working.

- Since we now have themes with different background colors I modified the `HIGHCOLOR` and `REL16` image drawing to replace color `0x0001` with the background color from the theme. None of the image files in `tft/images` uses `HIGHCOLOR` plus `0x0001` for any pixel color so this will only affect the `bootscreen_*.cpp` files where it was applied. We can add a note about this quirk somewhere. Since all the themes define a dark background this will not be noticeable if some future `HIGHCOLOR` bitmap ends up with `0x0001` pixels.

- I added a simple 240 x 250 bootscreen image for use with the 320 x 240 screen in portrait orientation.
  
  One strange thing is that the 240x250 boot screen is being partially repeated. It's not due to the image data being iterated more than once, so it must be something about the way the TFT screen buffer works. Maybe the buffer needs to be cleaned after the image is done drawing. You can see this in the Simulator using `TFT_RES_320x240` with `TFT_ROTATION TFT_ROTATE_90` and disabling `BOOT_MARLIN_LOGO_SMALL`. Interestingly, the same thing doesn't happen when using `BOOT_MARLIN_LOGO_SMALL` even though that uses the 195x59 image, which is also smaller than the screen.

**Comment by thinkyhead** [comment]
Aha. All boot screens that are smaller than the screen need to clear the background or else the TFT DMA will just emit whatever is in its buffer for the rest of the canvas. When `BOOT_MARLIN_LOGO_SMALL` is disabled, it no longer always means full-screen boot image, just the largest one. So with that correction the 240x320 boot image now displays correctly. Now this is good to go.

**Comment by ellensp** [comment]
As it was sitting on my bench I just tested on BOARD_MKS_ROBIN_NANO with MKS_ROBIN_TFT35 // 480x320, 3.5", FSMC Display From MKS
Works no issues. But is a bit slower to load than I would have expected
You can see it filling the screen..

Perhaps some optimization is in order.. (at some future time)
Or, quite likely, my expectations are just set to high...  It does have to decompress the data...

**Comment by Foxies-CSTL** [comment]
> As it was sitting on my bench I just tested on BOARD_MKS_ROBIN_NANO with MKS_ROBIN_TFT35 // 480x320, 3.5", FSMC Display From MKS Works no issues. But is a bit slower to load than I would have expected You can see it filling the screen..
> 
> Perhaps some optimization is in order.. (at some future time) Or, quite likely, my expectations are just set to high... It does have to decompress the data...

@ellensp, I also use a TFT-ROBIN-TFT35 but in portrait mode and the boot image overflows the screen. So I created a new image "MarlinLogo320x250x16" as @thinkyhead  did for 320x240 screens and declared it in "tft_image.h".

I also modified lines 59-60 of the "ui_320x480.h" file to recenter my custom logo (320x250).
It's an example:
image

So to summarize, here are the tests performed on physical hardware: 
- Mks_Robin_TFT32 (FMSC)= portrait/landscape mode ok
- Mks_Robin_TFT35 (FMSC)= portrait/landscape mode ok
- Mks_TS35R_V2 (SPI)= portrait/landscape mode ok


[bootscreen_320x250x16.zip](https://github.com/MarlinFirmware/Marlin/files/12171244/bootscreen_320x250x16.zip)

**Comment by jmz52** [comment]
> I also modified lines 59-60 of the "ui_320x480.h" file to recenter my custom logo (320x250).

@Foxies-CSTL, this is what `TFT_THEME` is for.
If you need custom layout, colors or bootscreen image - you should use custom theme for it instead of modifying the default layout.
Also your image is not full screen, so it should be used with `BOOT_MARLIN_LOGO_SMALL` option.

**Comment by Foxies-CSTL** [comment]
> @Foxies-CSTL, this is what `TFT_THEME` is for.

I've been using this principle for a long time and it's easier now with the Theme file, but here I'm talking about the splash image for a 480x320 screen in portrait mode that won't fit because its width is too big. Before putting my images and my colors, I make a good operation with the default parameters of the Bugfix sources.

> Also your image is not full screen, so it should be used with `BOOT_MARLIN_LOGO_SMALL` option.

Thanks, I know how it works ;-) .

**Comment by Foxies-CSTL** [comment]
For me it works fine but for bugfix users with a 480x320 screen in COLOR_UI portrait mode and without BOOT_MARLIN_LOGO_SMALL option, there will be a problem.

**Pull request merged**